source/common/upstream/upstream_impl.h

```diff
@@ -72,6 +72,38 @@ class HealthCheckHostMonitorNullImpl : public HealthCheckHostMonitor {
   void setUnhealthy(UnhealthyType) override {}
 };
 
+/**
+ * Implementation of LoadMetricStats.
+ */
+class LoadMetricStatsImpl : public LoadMetricStats {
+public:
+  LoadMetricStatsImpl() = default;
+
+  void add(const std::string& key, double value) override {
+    absl::MutexLock lock(&mu_);
+    if (map_ == nullptr) {
+      map_ = std::make_unique<StatsMap>();
+    }
+    Stat& stat = (*map_)[key];
+    ++stat.num_requests_with_metric;
+    stat.total_metric_value += value;
+  }
+
+  StatsMapPtr latch() override {
+    absl::MutexLock lock(&mu_);
+    StatsMapPtr latched = std::move(map_);
+    map_ = nullptr;
+    return latched;
+  }
+
+private:
+  absl::Mutex mu_;
+  StatsMapPtr map_ ABSL_GUARDED_BY(mu_) = nullptr;
```

Comment:
As discussed I think this of concern for memory use and lock contention. Can you add comments indicating this should be examined before this gets used in scenarios with large numbers of hosts?

Reply:
Done.